cEP0002.md

```diff
@@ -0,0 +1,166 @@
+
+A rough look at how the new local bear class implementation will look:
+
+```python
+class LocalBear:
+    def __init__(self, section, result_queue, file_dict):
+        self.section = section
+        self.result_queue = result_queue
+        self.file_dict = file_dict
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def __execute_task(self, task):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        for result in self.run(task, **self.kwargs):
+            self.result_queue.put(result)
+
+    def run(self,
+            file_proxy
+            *args,
+            dependency_results=None,
+            **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        for fileproxy in self.file_dict.values():
+            yield fileproxy
+```
+
+A test bear:
+
+```python
+class TestBear(LocalBear):
+    def run(self, file_proxy):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        for line, i in enumerate(file):
+            # This will contain the analysis routine
+            yield "Line has a problem" , file_proxy.name, i
+```
+
+A rough idea of how the new bear will be called:
+
+```python
+from multiprocessing import Pool, Queue
+from queue import Empty
+
+
+def all_results_ready(async_results):
+    """
+    Method checks if all bears have completed.
+    """
+    for result, i in reversed(enumerate(async_results)):
+        if result.ready():
+            del async_results[i]
+        else:
+            return False
+
+    return True
+
+
+def runstuff(handle_result):
+    queue = Queue()
+    bears = [LocalBear({}, queue, {})]
+
+    pool = Pool()
+    results = []
+    async_results = {}
+    for bear in bears:
+        # This is how the multiprocessing pool will call the bear.
+        async_results[bear] = pool.map_async(
+            bear.__execute_task,
+            bear.generate_tasks())
+
+    while True:
+        try:
+            result = queue.get(0.1)
+            if isinstance(result, Result):
+                handle_result(result)
+                results.append(result)
+            elif isinstance(result, LogMessage):
+                logger.log(result)
+        except Empty:
+            if all_results_ready(async_results):
+                break
+
+    return results
+```       break
+
+    return results
+```
```

Comment:
can you please read your own stuff before uploading this? This is probably even invalid markdown...